code/cabal-audit/src/Distribution/Audit.hs

```diff
@@ -0,0 +1,246 @@
+  HumanReadable -> humanReadableHandler . M.toList
+  Osv fp -> osvHandler fp
+
+osvHandler :: FilePath -> M.Map PackageName ElaboratedPackageInfoAdvised -> IO ()
+osvHandler fp =
+  Aeson.encodeFile @Value fp . object . M.foldMapWithKey
+    \pn MkElaboratedPackageInfoWith {elaboratedPackageVersion, packageAdvisories} ->
+      [ fromString (unPackageName pn)
+          .= object
+            [ "version" .= prettyVersion @Text elaboratedPackageVersion
+            , "advisories" .= map (OSV.convert . fst) (runIdentity packageAdvisories)
+            ]
+      ]
+
+-- | pretty-prints a `Version`
+--
+-- >>> import Distribution.Version
+-- >>> prettyVersion $ mkVersion [0, 1, 0, 0]
+-- "0.1.0.0"
+prettyVersion :: IsString s => Version -> s
+prettyVersion = fromString . List.intercalate "." . map show . versionNumbers
+{-# INLINE prettyVersion #-}
+
+prettyAdvisory :: Advisory -> Maybe Version -> Text
+prettyAdvisory Advisory {advisoryId, advisoryPublished, advisoryKeywords, advisorySummary} mfv =
+  T.unlines do
+    let hsecId = T.pack (printHsecId advisoryId)
+    map
+      ("  " <>)
+      [ formatWith [bold, blue] hsecId <> " \"" <> advisorySummary <> "\""
+      , "published: " <> formatWith [bold] (T.pack $ show advisoryPublished)
+      , "https://haskell.github.io/security-advisories/advisory/" <> hsecId
+      , fixAvailable
+      , formatWith [blue] $ T.intercalate ", " (coerce advisoryKeywords)
+      ]
+ where
+  fixAvailable = case mfv of
+    Nothing -> formatWith [bold, red] "No fix version available"
+    Just fv -> formatWith [bold, green] "Fix available since version " <> formatWith [yellow] (prettyVersion fv)
+
+-- | this is handler is used when displaying to the user
+humanReadableHandler :: [(PackageName, ElaboratedPackageInfoAdvised)] -> IO ()
+humanReadableHandler = \case
+  [] -> putStrLn (formatWith [green, bold] "No advisories found.")
+  avs -> do
+    putStrLn (formatWith [bold, red] "\n\nFound advisories:\n")
+    for_ avs \(pn, i) -> do
+      let verString = formatWith [yellow] $ prettyVersion $ elaboratedPackageVersion i
+          pkgName = formatWith [yellow] $ show $ unPackageName pn
+      putStrLn ("dependency " <> pkgName <> " at version " <> verString <> " is vulnerable for:")
+      for_ (runIdentity (packageAdvisories i)) (T.putStrLn . uncurry prettyAdvisory)
+
+projectConfigFromFlags :: NixStyleFlags a -> ProjectConfig
+projectConfigFromFlags flags = commandLineFlagsToProjectConfig defaultGlobalFlags flags mempty
+
+auditCommandParser :: Parser (AuditConfig, NixStyleFlags ())
+auditCommandParser =
+  (,)
+    <$> do
+      MkAuditConfig
+        <$> do
+          Left
+            <$> strOption do
+              mconcat
+                [ long "file-path"
+                , short 'p'
+                , metavar "FILEPATH"
+                , help "the path the the repository containing an advisories directory"
+                ]
+              <|> Right
+            <$> strOption do
```

Comment:
I suggest to build a proper URI parser. E.g. GHCup uses uri-bytestring:

* https://github.com/haskell/ghcup-hs/blob/be4a1bdf7a78f6783cba04de425b652d336192c3/lib-opt/GHCup/OptParse/Common.hs#L216-L217
* https://hackage.haskell.org/package/uri-bytestring-0.3.3.1/docs/URI-ByteString.html#v:parseURI

But network-uri probably does the job too:

* https://hackage.haskell.org/package/network-uri-2.6.4.2/docs/Network-URI.html#v:parseURI

Reply:
what would be the immediate advantage of that? I don't want to pass in my files as `file://` etc.

Reply:
> what would be the immediate advantage of that? I don't want to pass in my files as `file://` etc.

I'm not talking about filepath, but the URI options (for the git repo). Users passing malformed URIs will get errors once your code tries to git clone instead of directly.

Reply:
yeah, this assumes that you will pass a proper `git clone`eable `URI`, I don't know if the advantage of proper parsing and writing a bunch of handlers outweigh the advantage of the simplicity of this, for now? I can add this as future work to the tracking issue, though, if that's okay with you?